notes-0407-composition-structure.tex

```diff
@@ -0,0 +1,303 @@
+
+\subsection{Sigma}
+
+\begin{prooftree*}
+	\hypo{\oftype{M}{\dpairtype{x}{A}{B}}}
+	\hypo{\oftype{\varphi,\oftype{i}{\mathbb{I}}}{N}{\dpairtype{x}{A}{B}}}
+	\infer2{\eqterm{\mathtt{hcomp}^i\,\parens{\dpairtype{x}{A}{B}}\,[\varphi\mapsto N]\,M}
+	               {\pair{\mathtt{compFst}}{\mathtt{compSnd}}}
+	               {\dpairtype{x}{A}{B}}}
+\end{prooftree*}
+
+where
+
+\begin{align*}
+	\mathtt{compFst} & \defeq \mathtt{hcomp}^i\,A\,[\varphi\mapsto\pairfst{N}]\,\parens{\pairfst{M}} \\
+	\mathtt{compSnd} & \defeq \mathtt{comp}^i\,B\,[\varphi\mapsto\pairsnd{N}, x\mapsto\mathtt{filler}^i]\,\parens{\pairsnd{M}} \\
+	\mathtt{filler}^i & \defeq \mathtt{hfill}^i\,A\,[\varphi\mapsto\pairfst{N}]\,\parens{\pairfst{M}}
+\end{align*}
+
+\begin{prooftree*}
+	\hypo{\oftype{M}{\dpairtype{x}{A}{B}}}
+	\hypo{\oftype{r}{\mathbb{I}}}
+	\infer2{\eqterm{\mathtt{transp}^i\,\parens{\dpairtype{x}{A}{B}}\,\varphi\,M}
+	               {\pair{\mathtt{transpFst}}{\mathtt{transpSnd}}}
+	               {\dpairtype{x}{A}{B}}}
+\end{prooftree*}
+
+where
+
+\begin{align*}
+	\mathtt{transpFst} & \defeq \mathtt{transp}^i\,A\,\varphi\,\parens{\pairfst{M}} \\
+	\mathtt{transpSnd} & \defeq \mathtt{transp}^i\,\parens{B[x\mapsto\mathtt{filler}^i]}\,\varphi\,\parens{\pairsnd{M}} \\
+	\mathtt{filler}^i & \defeq \text{\texttt{transp-fill}}^i\,A\,\varphi\,\parens{\pairfst{M}}
+\end{align*}
+
+\subsection{Paths}
+
+\begin{prooftree*}
+	\hypo{\oftype{P}{\pathtype{j.A}{M}{N}}}
+	\hypo{\oftype{\varphi,\oftype{i}{\mathbb{I}}}{Q}{\pathtype{j.A}{M}{N}}}
+	\infer2{\eqterm{\mathtt{hcomp}^i\,\pathtype{j.A}{M}{N}\,[\varphi\mapsto Q]\,P}
+	               {\lam{j}{\mathtt{body}}}
+	               {\pathtype{j.A}{M}{N}}}
+\end{prooftree*}
+
+where
+
+\begin{align*}
+	\mathtt{body} & \defeq \mathtt{hcomp}^i\,A\,[\varphi\mapsto Q@j, j=0\mapsto M, j=1\mapsto N]\,\parens{P@j}
+\end{align*}
+
+\begin{prooftree*}
+	\hypo{\oftype{P}{\pathtype{j.A}{M}{N}}}
+	\infer1{\eqterm{\mathtt{transp}^i\,\parens{\pathtype{j.A}{M}{N}}\,\varphi\,P}
+	               {\lam{j}{\mathtt{body}}}
+	               {\pathtype{j.A}{M}{N}}}
+\end{prooftree*}
+
+where
+
+\begin{align*}
+	\mathtt{body} & \defeq \mathtt{comp}^i\,[\varphi\mapsto P@j, j=0\mapsto M, j=1\mapsto N]\,\parens{P@j}
+\end{align*}
+
+\subsection{Positive Types}
+
+Coercion simply ``just works'' for inductive types:
+
+\begin{prooftree*}
+	\hypo{A\,\text{is an inductive type}}
+	\hypo{\oftype{M}{A}}
+	\infer2{\eqterm{\mathtt{transp}^i\,A\,\varphi\,M}{M}{A}}
+\end{prooftree*}
+
+Homogeneous composition, on the other hand, requires a bit of a trick; since $N$ might not be constant (i.e., $N[i\mapsto0]\equiv N[i\mapsto1]$ might not hold). Here, $\mathtt{hcomp}$ adds additional values to the type; since type declarations only define an inductive type that's generated by a certain description, this is acceptable. Furthermore, the new values commute with elimination, which makes these new values aren't really ``stuck terms.''
```

Comment:
The reason is not because N is not constant (it *is* constant). The reason is that even if there is a path between two elements of N under some context, they are not provably judgmentally equal in our setting.